File: crates/workspaces/src/adapters/menus.m
#import <AppKit/AppKit.h>
#import <ApplicationServices/ApplicationServices.h>

typedef void (*neosicht_menu_cb)(void *, const char *);
typedef void (*neosicht_separator_cb)(void *);
typedef void (*neosicht_item_cb)(void *, const char *, const char *, bool, bool);

static NSString *stringAttribute(AXUIElementRef element, CFStringRef attribute) {
    CFTypeRef value = NULL;
    if (AXUIElementCopyAttributeValue(element, attribute, &value) != kAXErrorSuccess || value == NULL) return nil;
    if (CFGetTypeID(value) != CFStringGetTypeID()) { CFRelease(value); return nil; }
    return CFBridgingRelease(value);
}

static bool boolAttribute(AXUIElementRef element, CFStringRef attribute, bool fallback) {
    CFTypeRef value = NULL;
    if (AXUIElementCopyAttributeValue(element, attribute, &value) != kAXErrorSuccess || value == NULL) return fallback;
    bool result = CFGetTypeID(value) == CFBooleanGetTypeID() ? CFBooleanGetValue(value) : fallback;
    CFRelease(value);
    return result;
}

static NSArray *children(AXUIElementRef element) {
    CFTypeRef value = NULL;
    if (AXUIElementCopyAttributeValue(element, kAXChildrenAttribute, &value) != kAXErrorSuccess || value == NULL) return @[];
    if (CFGetTypeID(value) != CFArrayGetTypeID()) { CFRelease(value); return @[]; }
    return CFBridgingRelease(value);
}

static AXUIElementRef menuBarForApplication(const char *expectedName) {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        AXIsProcessTrustedWithOptions((__bridge CFDictionaryRef)@{
            (__bridge NSString *)kAXTrustedCheckOptionPrompt: @YES
        });
    });
    if (!AXIsProcessTrusted()) return NULL;
    NSRunningApplication *app = NSWorkspace.sharedWorkspace.frontmostApplication;
    if (app == nil) return NULL;
    NSString *expected = [NSString stringWithUTF8String:expectedName];
    if (expected.length > 0 && ![app.localizedName isEqualToString:expected]) return NULL;
    AXUIElementRef application = AXUIElementCreateApplication(app.processIdentifier);
    CFTypeRef menuBar = NULL;
    AXError error = AXUIElementCopyAttributeValue(application, kAXMenuBarAttribute, &menuBar);
    CFRelease(application);
    return error == kAXErrorSuccess ? menuBar : NULL;
}

static NSString *shortcutFor(AXUIElementRef item) {
    NSString *character = stringAttribute(item, kAXMenuItemCmdCharAttribute);
    if (character.length == 0) return nil;
    CFTypeRef modifiersValue = NULL;
    AXUIElementCopyAttributeValue(item, kAXMenuItemCmdModifiersAttribute, &modifiersValue);
    NSInteger modifiers = 0;
    if (modifiersValue != NULL && CFGetTypeID(modifiersValue) == CFNumberGetTypeID()) CFNumberGetValue(modifiersValue, kCFNumberNSIntegerType, &modifiers);
    if (modifiersValue != NULL) CFRelease(modifiersValue);
    NSMutableString *shortcut = [NSMutableString string];
    if (modifiers & kAXMenuItemModifierControl) [shortcut appendString:@"⌃"];
    if (modifiers & kAXMenuItemModifierOption) [shortcut appendString:@"⌥"];
    if (modifiers & kAXMenuItemModifierShift) [shortcut appendString:@"⇧"];
    if (!(modifiers & kAXMenuItemModifierNoCommand)) [shortcut appendString:@"⌘"];
    [shortcut appendString:character.uppercaseString];
    return shortcut;
}

bool neosicht_copy_frontmost_menus(const char *appName, void *context, neosicht_menu_cb menuCallback, neosicht_separator_cb separatorCallback, neosicht_item_cb itemCallback) {
    @autoreleasepool {
        AXUIElementRef menuBar = menuBarForApplication(appName);
        if (menuBar == NULL) return false;
        for (id topObject in children(menuBar)) {
            AXUIElementRef top = (__bridge AXUIElementRef)topObject;
            NSString *title = stringAttribute(top, kAXTitleAttribute);
            if (title.length == 0) continue;
            menuCallback(context, title.UTF8String);
            NSArray *topChildren = children(top);
            if (topChildren.count == 0) continue;
            AXUIElementRef menu = (__bridge AXUIElementRef)topChildren.firstObject;
            for (id itemObject in children(menu)) {
                AXUIElementRef item = (__bridge AXUIElementRef)itemObject;
                NSString *role = stringAttribute(item, kAXRoleAttribute);
                NSString *label = stringAttribute(item, kAXTitleAttribute);
                if ([role isEqualToString:(__bridge NSString *)kAXMenuItemRole] && label.length == 0) {
                    separatorCallback(context);
                    continue;
                }
                if (label.length == 0) continue;
                NSString *shortcut = shortcutFor(item);
                CFTypeRef markValue = NULL;
                bool checked = false;
                if (AXUIElementCopyAttributeValue(item, kAXMenuItemMarkCharAttribute, &markValue) == kAXErrorSuccess && markValue != NULL) {
                    checked = CFGetTypeID(markValue) == CFStringGetTypeID() && [(__bridge NSString *)markValue length] > 0;
                    CFRelease(markValue);
                }
                itemCallback(context, label.UTF8String, shortcut.UTF8String, checked, boolAttribute(item, kAXEnabledAttribute, true));
            }
        }
        CFRelease(menuBar);
        return true;
    }
}

bool neosicht_activate_frontmost_menu_item(const char *appName, const char *menuTitle, size_t itemIndex) {
    @autoreleasepool {
        AXUIElementRef menuBar = menuBarForApplication(appName);
        if (menuBar == NULL) return false;
        NSString *wantedMenu = [NSString stringWithUTF8String:menuTitle];
        bool activated = false;
        for (id topObject in children(menuBar)) {
            AXUIElementRef top = (__bridge AXUIElementRef)topObject;
            if (![stringAttribute(top, kAXTitleAttribute) isEqualToString:wantedMenu]) continue;
            NSArray *topChildren = children(top);
            if (topChildren.count == 0) break;
            AXUIElementRef menu = (__bridge AXUIElementRef)topChildren.firstObject;
            size_t entryIndex = 0;
            for (id itemObject in children(menu)) {
                AXUIElementRef item = (__bridge AXUIElementRef)itemObject;
                NSString *role = stringAttribute(item, kAXRoleAttribute);
                NSString *label = stringAttribute(item, kAXTitleAttribute);
                bool isSeparator = [role isEqualToString:(__bridge NSString *)kAXMenuItemRole]
                    && label.length == 0;
                if (!isSeparator && label.length == 0) continue;
                if (entryIndex == itemIndex) {
                    activated = !isSeparator
                        && AXUIElementPerformAction(item, kAXPressAction) == kAXErrorSuccess;
                    break;
                }
                entryIndex += 1;
            }
            break;
        }
        CFRelease(menuBar);
        return activated;
    }
}

File: crates/workspaces/src/adapters/menus.swift
import AppKit
import ApplicationServices
import Foundation

public typealias MenuCallback = @convention(c) (UnsafeMutableRawPointer?, UnsafePointer<CChar>?) -> Void
public typealias SeparatorCallback = @convention(c) (UnsafeMutableRawPointer?) -> Void
public typealias ItemCallback = @convention(c) (
    UnsafeMutableRawPointer?, UnsafePointer<CChar>?, UnsafePointer<CChar>?, Bool, Bool
) -> Void

private func stringAttribute(_ element: AXUIElement, _ attribute: CFString) -> String? {
    var value: CFTypeRef?
    guard AXUIElementCopyAttributeValue(element, attribute, &value) == .success else { return nil }
    return value as? String
}

private func boolAttribute(_ element: AXUIElement, _ attribute: CFString, fallback: Bool) -> Bool {
    var value: CFTypeRef?
    guard AXUIElementCopyAttributeValue(element, attribute, &value) == .success else { return fallback }
    return (value as? Bool) ?? fallback
}

private func children(_ element: AXUIElement) -> [AXUIElement] {
    var value: CFTypeRef?
    guard AXUIElementCopyAttributeValue(element, kAXChildrenAttribute as CFString, &value) == .success,
          let array = value as? [AXUIElement]
    else { return [] }
    return array
}

private var permissionPrompted = false
private func menuBar(expectedName: String) -> AXUIElement? {
    if !permissionPrompted {
        permissionPrompted = true
        AXIsProcessTrustedWithOptions([kAXTrustedCheckOptionPrompt.takeUnretainedValue() as String: true] as CFDictionary)
    }
    guard AXIsProcessTrusted(),
          let app = NSWorkspace.shared.frontmostApplication,
          expectedName.isEmpty || app.localizedName == expectedName
    else { return nil }
    let application = AXUIElementCreateApplication(app.processIdentifier)
    var value: CFTypeRef?
    guard AXUIElementCopyAttributeValue(application, kAXMenuBarAttribute as CFString, &value) == .success else {
        return nil
    }
    return (value as! AXUIElement)
}

private func shortcut(_ item: AXUIElement) -> String? {
    guard let character = stringAttribute(item, kAXMenuItemCmdCharAttribute as CFString),
          !character.isEmpty else { return nil }
    var value: CFTypeRef?
    var modifiers = 0
    if AXUIElementCopyAttributeValue(item, kAXMenuItemCmdModifiersAttribute as CFString, &value) == .success,
       let number = value as? NSNumber {
        modifiers = number.intValue
    }
    // AX modifier constants are C enum values not imported by Swift.
    let shift = 1 << 0
    let option = 1 << 1
    let control = 1 << 2
    let noCommand = 1 << 3
    var result = ""
    if modifiers & control != 0 { result += "⌃" }
    if modifiers & option != 0 { result += "⌥" }
    if modifiers & shift != 0 { result += "⇧" }
    if modifiers & noCommand == 0 { result += "⌘" }
    return result + character.uppercased()
}

private func withCString(_ value: String?, _ body: (UnsafePointer<CChar>?) -> Void) {
    guard let value else { body(nil); return }
    value.withCString(body)
}

@_cdecl("neosicht_copy_frontmost_menus")
public func neosichtCopyFrontmostMenus(
    _ appName: UnsafePointer<CChar>?,
    _ context: UnsafeMutableRawPointer?,
    _ menuCallback: MenuCallback,
    _ separatorCallback: SeparatorCallback,
    _ itemCallback: ItemCallback
) -> Bool {
    let expected = appName.map(String.init(cString:)) ?? ""
    guard let bar = menuBar(expectedName: expected) else { return false }
    for top in children(bar) {
        guard let title = stringAttribute(top, kAXTitleAttribute as CFString), !title.isEmpty else { continue }
        title.withCString { menuCallback(context, $0) }
        guard let menu = children(top).first else { continue }
        for item in children(menu) {
            let role = stringAttribute(item, kAXRoleAttribute as CFString)
            let label = stringAttribute(item, kAXTitleAttribute as CFString) ?? ""
            let separator = role == (kAXMenuItemRole as String) && label.isEmpty
            if separator { separatorCallback(context); continue }
            if label.isEmpty { continue }
            let marked = !(stringAttribute(item, kAXMenuItemMarkCharAttribute as CFString) ?? "").isEmpty
            let enabled = boolAttribute(item, kAXEnabledAttribute as CFString, fallback: true)
            label.withCString { labelPointer in
                withCString(shortcut(item)) { shortcutPointer in
                    itemCallback(context, labelPointer, shortcutPointer, marked, enabled)
                }
            }
        }
    }
    return true
}

@_cdecl("neosicht_activate_frontmost_menu_item")
public func neosichtActivateFrontmostMenuItem(
    _ appName: UnsafePointer<CChar>?,
    _ menuTitle: UnsafePointer<CChar>?,
    _ itemIndex: Int
) -> Bool {
    let expected = appName.map(String.init(cString:)) ?? ""
    let wanted = menuTitle.map(String.init(cString:)) ?? ""
    guard let bar = menuBar(expectedName: expected) else { return false }
    for top in children(bar) where stringAttribute(top, kAXTitleAttribute as CFString) == wanted {
        guard let menu = children(top).first else { return false }
        var entryIndex = 0
        for item in children(menu) {
            let role = stringAttribute(item, kAXRoleAttribute as CFString)
            let label = stringAttribute(item, kAXTitleAttribute as CFString) ?? ""
            let separator = role == (kAXMenuItemRole as String) && label.isEmpty
            if !separator && label.isEmpty { continue }
            if entryIndex == itemIndex {
                return !separator && AXUIElementPerformAction(item, kAXPressAction as CFString) == .success
            }
            entryIndex += 1
        }
    }
    return false
}
